Could not compile Elmer with VTK 9.0.3.

Hi,

I packaging ElmerGUI as Flatpak for Linux.
Flatpak is a way to package Linux software that include any dependencies the app needs. Which mean package the app once and work on all distro.

Here is what I have so far.
https://github.com/kevinsmia1939/flathub/tree/fi.csc.Elmer

I want to support many feature as possible, but I have some issues.

It look like Elmer does not compile with VTK 9.0.3.
The flags I use are listed below.

```
      - -DBUILD_SHARED_LIBS=ON
      - -DOpenGL_GL_PREFERENCE=GLVND
      - -DBUILD_TESTING:BOOL=OFF
      - -DVTK_USE_SYSTEM_ZLIB:BOOL=ON
      - -DVTK_USE_SYSTEM_EIGEN:BOOL=ON
      - -DVTK_BUILD_QT_DESIGNER_PLUGIN:BOOL=ON
      - -DVTK_GROUP_ENABLE_Qt=WANT
      - -DCMAKE_BUILD_TYPE=Release
      - -DVTK_LEGACY_REMOVE:BOOL=ON
      - -DVTK_GROUP_ENABLE_Rendering=WANT
      - -DVTK_GROUP_ENABLE_StandAlone=WANT
      - -DVTK_USE_MPI:BOOL=ON
      - -DVTK_GROUP_ENABLE_MPI=WANT
      - -DVTK_MODULE_ENABLE_VTK_CommonSystem=YES
      - -DVTK_MODULE_ENABLE_VTK_FiltersGeneral=YES
      - -DVTK_MODULE_ENABLE_VTK_FiltersGeometry=YES
      - -DVTK_MODULE_ENABLE_VTK_ImagingCore=YES
      - -DVTK_MODULE_ENABLE_VTK_ImagingHybrid=YES
      - -DVTK_MODULE_ENABLE_VTK_InteractionStyle=YES
      - -DVTK_MODULE_ENABLE_VTK_InteractionWidgets=YES
      - -DVTK_MODULE_ENABLE_VTK_IOCityGML=YES
      - -DVTK_MODULE_ENABLE_VTK_IOExodus=YES
      - -DVTK_MODULE_ENABLE_VTK_IOGeometry=YES
      - -DVTK_MODULE_ENABLE_VTK_IOImage=YES
      - -DVTK_MODULE_ENABLE_VTK_IOImport=YES
      - -DVTK_MODULE_ENABLE_VTK_IOParallel=YES
      - -DVTK_MODULE_ENABLE_VTK_IOPLY=YES
      - -DVTK_MODULE_ENABLE_VTK_IOXML=YES
      - -DVTK_MODULE_ENABLE_VTK_RenderingAnnotation=YES
      - -DVTK_MODULE_ENABLE_VTK_RenderingCore=YES
      - -DVTK_MODULE_ENABLE_VTK_RenderingLabel=YES
      - -DVTK_MODULE_ENABLE_VTK_RenderingOpenGL2=YES
      - -DVTK_MODULE_ENABLE_VTK_RenderingVolumeOpenGL2=YES
      - -DVTK_MODULE_ENABLE_VTK_jsoncpp=YES
```

I use latest Elmer in devel branch.

```
      - -DCMAKE_BUILD_TYPE:STRING=Release
      - -DWITH_ELMERGUI:BOOL=TRUE
      - -DWITH_ELMERGUILOGGER:BOOL=TRUE
      - -DWITH_CONTRIB=ON
      - -DWITH_OpenMP:BOOL=TRUE
      - -DWITH_MPI:BOOL=TRUE
      - -DWITH_Mumps:BOOL=TRUE
      - -DWITH_Hypre:BOOL=TRUE
      - -DWITH_ScatteredDataInterpolator=ON
      - -DWITH_ElmerIce:BOOL=TRUE
      - -DMumps_INCLUDE_DIR=/app/include
      - -DMumps_LIBRARIES=/app/lib
      - -DWITH_QT5=TRUE
      - -DWITH_VTK=ON
```

gcc 10.2.0
cmake 3.20.2

I use openSUSE Tumbleweed.
```
Operating System: openSUSE Tumbleweed 20210929
KDE Plasma Version: 5.22.5
KDE Frameworks Version: 5.86.0
Qt Version: 5.15.2
Kernel Version: 5.14.6-1-default (64-bit)
Graphics Platform: X11
Processors: 12 × AMD Ryzen 5 4600H with Radeon Graphics
Memory: 7.1 GiB of RAM
Graphics Processor: AMD RENOIR
```

```
[ 65%] Building Fortran object fem/src/CMakeFiles/elmersolver.dir/IterativeMethods.F90.o
/run/build/Elmer/ElmerGUI/Application/vtkpost/vtkpost.cpp:84:10: fatal error: QVTKWidget.h: No such file or directory
   84 | #include <QVTKWidget.h>
      |          ^~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [ElmerGUI/Application/CMakeFiles/ElmerGUI.dir/build.make:847: ElmerGUI/Application/CMakeFiles/ElmerGUI.dir/vtkpost/vtkpost.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 65%] Building Fortran object fem/src/CMakeFiles/elmersolver.dir/ListMatrix.F90.o
In file included from /run/build/Elmer/ElmerGUI/Application/src/mainwindow.h:63,
                 from /run/build/Elmer/ElmerGUI/Application/src/convergenceview.cpp:44:
/run/build/Elmer/ElmerGUI/Application/src/objectbrowser.h:52:60: warning: ‘constexpr QFlags<T>::QFlags(QFlags<T>::Zero) [with Enum = Qt::WindowType; QFlags<T>::Zero = int QFlags<Qt::WindowType>::Private::*]’ is deprecated: Use default constructor instead [-Wdeprecated-declarations]
   52 |   ObjectBrowser(QMainWindow *parent, Qt::WindowFlags flags=0);
      |                                                            ^
In file included from /usr/include/QtCore/qglobal.h:1299,
                 from /usr/include/QtCore/QtCore:4,
                 from /usr/include/QtGui/QtGuiDepends:3,
                 from /usr/include/QtGui/QtGui:3,
                 from /run/build/Elmer/ElmerGUI/Application/src/convergenceview.cpp:41:
/usr/include/QtCore/qflags.h:123:80: note: declared here
  123 |     QT_DEPRECATED_X("Use default constructor instead") Q_DECL_CONSTEXPR inline QFlags(Zero) noexcept : i(0) {}
      |                                                                                ^~~~~~
/run/build/Elmer/ElmerGUI/Application/src/convergenceview.cpp: In member function ‘void ConvergenceView::savePictureSlot()’:
/run/build/Elmer/ElmerGUI/Application/src/convergenceview.cpp:329:44: warning: ‘static QPixmap QPixmap::grabWidget(QObject*, int, int, int, int)’ is deprecated: Use QWidget::grab() instead [-Wdeprecated-declarations]
  329 |   QPixmap pixmap = QPixmap::grabWidget(plot);
      |                                            ^
In file included from /usr/include/QtGui/qbrush.h:52,
                 from /usr/include/QtGui/qpen.h:45,
                 from /usr/include/QtGui/qtextformat.h:49,
                 from /usr/include/QtGui/qtextlayout.h:50,
                 from /usr/include/QtGui/qabstracttextdocumentlayout.h:45,
                 from /usr/include/QtGui/QtGui:5,
                 from /run/build/Elmer/ElmerGUI/Application/src/convergenceview.cpp:41:
/usr/include/QtGui/qpixmap.h:122:20: note: declared here
  122 |     static QPixmap grabWidget(QObject *widget, int x = 0, int y = 0, int w = -1, int h = -1);
      |                    ^~~~~~~~~~
/run/build/Elmer/ElmerGUI/Application/src/convergenceview.cpp:339:18: warning: ignoring return value of ‘QString QString::toUpper() const &’, declared with attribute ‘nodiscard’ [-Wunused-result]
  339 |   suffix.toUpper();
      |                  ^
In file included from /usr/include/QtCore/qobject.h:47,
                 from /usr/include/QtCore/qabstractanimation.h:43,
                 from /usr/include/QtCore/QtCore:6,
                 from /usr/include/QtGui/QtGuiDepends:3,
                 from /usr/include/QtGui/QtGui:3,
                 from /run/build/Elmer/ElmerGUI/Application/src/convergenceview.cpp:41:
/usr/include/QtCore/qstring.h:512:31: note: declared here
  512 |     Q_REQUIRED_RESULT QString toUpper() const &
      |                               ^~~~~~~
/run/build/Elmer/ElmerGUI/Application/vtkpost/featureedge.cpp: In member function ‘void FeatureEdge::draw(VtkPost*, Preferences*)’:
/run/build/Elmer/ElmerGUI/Application/vtkpost/featureedge.cpp:99:11: error: ‘class vtkGeometryFilter’ has no member named ‘SetInput’; did you mean ‘GetInput’?
   99 |   filter->SetInput(grid);
      |           ^~~~~~~~
      |           GetInput
[ 65%] Building Fortran object fem/src/CMakeFiles/elmersolver.dir/IterSolve.F90.o
/run/build/Elmer/ElmerGUI/Application/vtkpost/isosurface.cpp: In member function ‘void IsoSurface::draw(VtkPost*, TimeStep*)’:
/run/build/Elmer/ElmerGUI/Application/vtkpost/isosurface.cpp:181:21: warning: ‘void qSort(Container&) [with Container = QVector<double>]’ is deprecated: Use std::sort [-Wdeprecated-declarations]
  181 |   qSort(contourValue);
      |                     ^
In file included from /usr/include/QtCore/qlist.h:43,
                 from /usr/include/QtCore/qobject.h:49,
                 from /usr/include/QtCore/qabstractanimation.h:43,
                 from /usr/include/QtCore/QtCore:6,
                 from /usr/include/QtGui/QtGuiDepends:3,
                 from /usr/include/QtGui/QtGui:3,
                 from /run/build/Elmer/ElmerGUI/Application/vtkpost/isosurface.cpp:41:
/usr/include/QtCore/qalgorithms.h:188:46: note: declared here
  188 | QT_DEPRECATED_X("Use std::sort") inline void qSort(Container &c)
      |                                              ^~~~~
/run/build/Elmer/ElmerGUI/Application/vtkpost/isosurface.cpp:230:8: error: ‘class vtkContourFilter’ has no member named ‘SetInput’; did you mean ‘GetInput’?
  230 |   iso->SetInput(vtkPost->GetVolumeGrid());
      |        ^~~~~~~~
      |        GetInput

```

Thanks.

--- a/ElmerGUI/Application/cad/cadview.cpp
+++ b/ElmerGUI/Application/cad/cadview.cpp
@@ -150,7 +150,11 @@ CadView::CadView(QWidget *parent) : QMainWindow(parent) {
 
   renderer = vtkRenderer::New();
   renderer->SetBackground(1, 1, 1);
+#if VTK_MAJOR_VERSION >=9
+  qVTKWidget->renderWindow()->AddRenderer(renderer);
+#else
   qVTKWidget->GetRenderWindow()->AddRenderer(renderer);
+#endif
   renderer->GetRenderWindow()->Render();
 
   vtkPropPicker *propPicker = vtkPropPicker::New();
@@ -502,7 +506,11 @@ bool CadView::readFile(QString fileName) {
   // Draw:
   //------
   renderer->ResetCamera();  
+#if VTK_MAJOR_VERSION >= 9
+  qVTKWidget->renderWindow()->Render();
+#else
   qVTKWidget->GetRenderWindow()->Render();
+#endif
 
   QCoreApplication::processEvents();
 

--- a/ElmerGUI/Application/cad/cadview.cpp
+++ b/ElmerGUI/Application/cad/cadview.cpp
@@ -104,7 +104,11 @@ static void pickEventHandler(vtkObject* caller, unsigned long eid,
   QVTKWidget* qvtkWidget = cadView->GetQVTKWidget();
 #endif
 
+#if VTK_MAJOR_VERSION >= 9
+  vtkAbstractPicker* picker = qvtkWidget->interactor()->GetPicker();
+#else
   vtkAbstractPicker* picker = qvtkWidget->GetInteractor()->GetPicker();
+#endif
   vtkPropPicker* propPicker = vtkPropPicker::SafeDownCast(picker);
   vtkActor* actor = propPicker->GetActor();
 
@@ -151,11 +155,19 @@ CadView::CadView(QWidget *parent) : QMainWindow(parent) {
 
   vtkPropPicker *propPicker = vtkPropPicker::New();
   vtkCallbackCommand *cbcPick = vtkCallbackCommand::New();
+#if VTK_MAJOR_VERSION >= 9
+  qVTKWidget->interactor()->SetPicker(propPicker);
+  cbcPick->SetClientData(this);
+  cbcPick->SetCallback(pickEventHandler);
+  qVTKWidget->interactor()->GetPicker()->AddObserver(vtkCommand::PickEvent,
+                                                        cbcPick);
+#else
   qVTKWidget->GetInteractor()->SetPicker(propPicker);
   cbcPick->SetClientData(this);
   cbcPick->SetCallback(pickEventHandler);
   qVTKWidget->GetInteractor()->GetPicker()->AddObserver(vtkCommand::PickEvent,
                                                         cbcPick);
+#endif
   propPicker->Delete();
   cbcPick->Delete();
 

--- a/ElmerGUI/Application/cad/cadview.h
+++ b/ElmerGUI/Application/cad/cadview.h
@@ -58,9 +58,8 @@ namespace nglib {
 #include <BRepMesh_IncrementalMesh.hxx>
 #endif
 
-#if 1
 #include "vtkConfigure.h"
-#else
+#ifndef VTK_MAJOR_VERSION
 #include "vtkVersionMacros.h"
 #endif
 

--- a/ElmerGUI/Application/vtkpost/vtkpost.cpp
+++ b/ElmerGUI/Application/vtkpost/vtkpost.cpp
@@ -171,7 +171,11 @@ static void pEventHandler(vtkObject* caller, unsigned long eid,
   QVTKWidget* qvtkWidget = vtkPost->GetQVTKWidget();
 #endif
 
+#if VTK_MAJOR_VERSION >= 9
+  vtkAbstractPicker* picker = qvtkWidget->interactor()->GetPicker();
+#else
   vtkAbstractPicker* picker = qvtkWidget->GetInteractor()->GetPicker();
+#endif
   vtkPropPicker* propPicker = vtkPropPicker::SafeDownCast(picker);
 
   vtkActor* actor = propPicker->GetActor();
@@ -382,14 +386,22 @@ VtkPost::VtkPost(QWidget *parent)
   // Create a cell picker and set the callback & observer:
   //------------------------------------------------------
   vtkPropPicker* propPicker = vtkPropPicker::New();
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->interactor()->SetPicker(propPicker);
+#else
   qvtkWidget->GetInteractor()->SetPicker(propPicker);
+#endif
   propPicker->Delete();
 
   vtkCallbackCommand* cbcPick = vtkCallbackCommand::New();
   cbcPick->SetClientData(this);
   cbcPick->SetCallback(pEventHandler);
 
+#if VTK_MAJOR_VERSION >= 9
+  vtkAbstractPicker* picker = qvtkWidget->interactor()->GetPicker();
+#else
   vtkAbstractPicker* picker = qvtkWidget->GetInteractor()->GetPicker();
+#endif
   picker->AddObserver(vtkCommand::EndPickEvent, cbcPick);
   cbcPick->Delete();
 
@@ -402,7 +414,11 @@ VtkPost::VtkPost(QWidget *parent)
   cbcPlane->SetCallback(iEventHandler);
 
   planeWidget = vtkImplicitPlaneWidget::New();
+#if VTK_MAJOR_VERSION >= 9
+  planeWidget->SetInteractor(qvtkWidget->interactor());
+#else
   planeWidget->SetInteractor(qvtkWidget->GetInteractor());
+#endif
   planeWidget->AddObserver(vtkCommand::InteractionEvent, cbcPlane);
   cbcPlane->Delete();
 

--- a/ElmerGUI/Application/vtkpost/vtkpost.cpp
+++ b/ElmerGUI/Application/vtkpost/vtkpost.cpp
@@ -380,7 +380,11 @@ VtkPost::VtkPost(QWidget *parent)
   //------------------
   renderer = vtkRenderer::New();
   renderer->SetBackground(1, 1, 1);
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->AddRenderer(renderer);
+#else
   qvtkWidget->GetRenderWindow()->AddRenderer(renderer);
+#endif
   renderer->GetRenderWindow()->Render();
 
   // Create a cell picker and set the callback & observer:
@@ -2216,7 +2220,11 @@ void VtkPost::redrawSlot()
   drawAxesSlot();
   drawTextSlot();
 
+#if VTK_MAJOR_VERSION >= 9
+  vtkRenderWindow *renderWindow = qvtkWidget->renderWindow();
+#else
   vtkRenderWindow *renderWindow = qvtkWidget->GetRenderWindow();
+#endif
   renderWindow->Render();
 
   // Check if the "Stop" button of time stepping loop has been pressed:
@@ -2234,7 +2242,11 @@ void VtkPost::redrawSlot()
 void VtkPost::showColorBarDialogSlot()
 {
   if(!postFileRead) return;
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 
   if(drawColorBarAct->isChecked()) {
     colorBar->show();
@@ -2257,7 +2269,11 @@ void VtkPost::drawColorBarSlot()
   if(!drawColorBarAct->isChecked()) return;
   colorBar->draw(this);
   renderer->AddActor(colorBarActor);
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw mesh points:
@@ -2269,7 +2285,11 @@ void VtkPost::drawMeshPointSlot()
   if(!drawMeshPointAct->isChecked()) return;
   meshPoint->draw(this, preferences);
   renderer->AddActor(meshPointActor);
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw mesh edges:
@@ -2281,7 +2301,11 @@ void VtkPost::drawMeshEdgeSlot()
   if(!drawMeshEdgeAct->isChecked()) return;
   meshEdge->draw(this, preferences);
   renderer->AddActor(meshEdgeActor);
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw feature edges:
@@ -2293,7 +2317,11 @@ void VtkPost::drawFeatureEdgesSlot()
   if(!drawFeatureEdgesAct->isChecked()) return;
   featureEdge->draw(this, preferences);
   renderer->AddActor(featureEdgeActor);
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw stream lines:
@@ -2301,7 +2329,11 @@ void VtkPost::drawFeatureEdgesSlot()
 void VtkPost::showStreamLineDialogSlot()
 {
   if(!postFileRead) return;
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
   
   if(drawStreamLineAct->isChecked()) {
     streamLine->show();
@@ -2325,7 +2357,11 @@ void VtkPost::drawStreamLineSlot()
   streamLine->draw(this, timeStep);
   renderer->AddActor(streamLineActor);
   drawColorBarSlot();
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw vectors:
@@ -2333,7 +2369,11 @@ void VtkPost::drawStreamLineSlot()
 void VtkPost::showVectorDialogSlot()
 {
   if(!postFileRead) return;
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 
   if(drawVectorAct->isChecked()) {
     vector->show();
@@ -2357,7 +2397,11 @@ void VtkPost::drawVectorSlot()
   vector->draw(this, timeStep);
   renderer->AddActor(vectorActor);
   drawColorBarSlot();
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw surfaces:
@@ -2365,7 +2409,11 @@ void VtkPost::drawVectorSlot()
 void VtkPost::showSurfaceDialogSlot()
 {
   if(!postFileRead) return;
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 
   if(drawSurfaceAct->isChecked()) {
     surface->show();
@@ -2389,7 +2437,11 @@ void VtkPost::drawSurfaceSlot()
   surface->draw(this, timeStep);
   renderer->AddActor(surfaceActor);
   drawColorBarSlot();
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw iso contours (2D):
@@ -2397,7 +2449,11 @@ void VtkPost::drawSurfaceSlot()
 void VtkPost::showIsoContourDialogSlot()
 {
   if(!postFileRead) return;
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 
   if(drawIsoContourAct->isChecked()) {
     isoContour->show();
@@ -2421,7 +2477,11 @@ void VtkPost::drawIsoContourSlot()
   isoContour->draw(this, timeStep);
   renderer->AddActor(isoContourActor);
   drawColorBarSlot();  
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw isosurfaces (3D):
@@ -2429,7 +2489,11 @@ void VtkPost::drawIsoContourSlot()
 void VtkPost::showIsoSurfaceDialogSlot()
 {
   if(!postFileRead) return;
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 
   if(drawIsoSurfaceAct->isChecked()) {
     isoSurface->show();
@@ -2453,7 +2517,11 @@ void VtkPost::drawIsoSurfaceSlot()
   isoSurface->draw(this, timeStep);
   renderer->AddActor(isoSurfaceActor);
   drawColorBarSlot();
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw axes:
@@ -2471,7 +2539,11 @@ void VtkPost::drawAxesSlot()
   renderer->AddActor(axesXTextActor);
   renderer->AddActor(axesYTextActor);
   renderer->AddActor(axesZTextActor);
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 // Draw text:
@@ -2479,7 +2551,11 @@ void VtkPost::drawAxesSlot()
 void VtkPost::showTextDialogSlot()
 {
   if(!postFileRead) return;
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 
   if(drawTextAct->isChecked()) {
     text->show();
@@ -2502,7 +2578,11 @@ void VtkPost::drawTextSlot()
   if(!drawTextAct->isChecked()) return;
   text->draw(this);
   renderer->AddActor2D(textActor);
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 
@@ -2865,7 +2945,11 @@ void VtkPost::SetPostFileEnd(int n)
 //------------------------------------------------------------
 void VtkPost::Render()
 {
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
 }
 
 void VtkPost::Redraw()
@@ -3354,7 +3438,11 @@ bool VtkPost::SavePngFile(QString fileName)
 
   vtkWindowToImageFilter* image = vtkWindowToImageFilter::New();
 
+#if VTK_MAJOR_VERSION >= 9
+  image->SetInput(qvtkWidget->renderWindow());
+#else
   image->SetInput(qvtkWidget->GetRenderWindow());
+#endif
   image->Update();
 
   vtkPNGWriter* writer = vtkPNGWriter::New();
@@ -3366,7 +3454,11 @@ bool VtkPost::SavePngFile(QString fileName)
 #else
   writer->SetFileName(fileName.toAscii().data());
 #endif
+#if VTK_MAJOR_VERSION >= 9
+  qvtkWidget->renderWindow()->Render();
+#else
   qvtkWidget->GetRenderWindow()->Render();
+#endif
   writer->Write();
 
   writer->Delete();

--- a/ElmerGUI/Application/vtkpost/vtkpost.h
+++ b/ElmerGUI/Application/vtkpost/vtkpost.h
@@ -57,9 +57,8 @@
 #include <gui/PythonQtScriptingConsole.h>
 #endif
 
-#if 1
 #include "vtkConfigure.h"
-#else
+#ifndef VTK_MAJOR_VERSION
 #include "vtkVersionMacros.h"
 #endif
 

--- a/fem/CMakeLists.txt
+++ b/fem/CMakeLists.txt
@@ -62,6 +62,8 @@ IF(APPLE)
   SET(SHL_EXTENSION ".dylib")
 ENDIF()
 
+# this should just use ELMER_FEM_VERSION as plane VERSION is to bread on prone to conflicts
+set(VERSION ${ELMER_FEM_VERSION})
 CONFIGURE_FILE(config.h.cmake config.h.commented)
 FORTRANIZE(${CMAKE_BINARY_DIR}/fem/config.h.commented ${CMAKE_BINARY_DIR}/fem/config.h)
 

